locator/tablets.cc

```diff
@@ -430,6 +430,31 @@ class tablet_effective_replication_map : public effective_replication_map {
         return result;
     }
 
+    std::optional<tablet_routing_info> check_locality(const token& search_token, unsigned shard) const override {
+        auto&& tablets = get_tablet_map();
+        auto tablet = tablets.get_tablet_id(search_token);
+        auto&& info = tablets.get_tablet_transition_info(tablet);
+        auto&& replicas = std::invoke([&] () -> const tablet_replica_set& {
+            if (!info) {
+                return tablets.get_tablet_info(tablet).replicas;
+            }
+            switch (info->reads) {
```

Comment:
Since this is used also for checking locality of writes, we cannot rely only on read selector. During migration, writes start to use the new replica before reads.

We should communicate the selector via check_locality() parameter to be precise. Or, we can assume that drivers can use all the replicas during migration. That would simplify the check a bit, while temporarily incurring overhead of an extra hop during migration post streaming for some requests. WDYT @avikivity.

Reply:
It actually makes sense to stick to a single replica set if drivers don't distinguish them between reads and writes. Perhaps some temporary mismatch is ok, especially that the windows are short in time.

Reply:
Can we communicate the subset that is available for both reads and writes?

Reply:
We can, but that complicates the driver side to manage 2 replica sets. Is it worth it to avoid temporary (post streaming) extra hop for a fraction of tablet requests, which is already a fraction of all requests since only a few tablets are migrating at a time?

Reply:
We have to decide what to do here. I think it's ok to use a single replica set.

Reply:
Note, if we send a subset and then the subset expands, we won't have an indication that we need to update the driver about the missing replica.

Reply:
> Note, if we send a subset and then the subset expands, we won't have an indication that we need to update the driver about the missing replica.

That seems like an orthogonal problem to the one mentioned above, you have it even if we split replica sets between reads and writes. To handle expansion, we can send events.

Reply:
yes, though I had hoped to only have one mechanism